### PR TITLE
[Auditbeat] Update host metricset

### PR DESCRIPTION
**Pull request by cwurm**
Similar to https://github.com/elastic/beats/pull/9139, this updates the `host` metricset to be in line with the other metricsets in the `system` module.

High-level changes:

1. Adds regular state reporting based on `state.period`/`host.state.period`
2. Persists state between restarts in `beat.db`
3. Detects changes in host information, with `event.action`:
3.1 `host_id_changed` (when `system.host.id` changes)
3.2 `reboot` (when `system.host.boottime` changes)
3.3 `host_changed` (for all other changes, e.g. hostname, IPs)
4. Changes to using `system.host.ip`/`system.host.mac` instead of `system.host.network.interfaces` - the `add_host_metadata` processor reports IPs and MACs only, too, and I don't see value at the moment of reporting the full network information. We can always add it later.

In contrast to the other metricsets, this one maintains its own `system.host` namespace instead of using the top-level `host`. This is because `host` is filled by the `add_host_metadata` processor. The processor uses cached values, and so a change event would be accompanied by unchanged data.

### CODE REVIEW

File: x-pack/auditbeat/module/system/host/_meta/data.json [hunk 1]
```diff
@@ -1,46 +1,51 @@
 {
     "@timestamp": "2017-10-12T08:05:34.853Z",
-    "beat": {
+    "agent": {
```
Comment (tsg) [inline x-pack/auditbeat/module/system/host/_meta/data.json]: This is not part of this PR, right? I'm asking because this change shouldn't get into 6.6.

File: x-pack/auditbeat/module/system/host/host.go [hunk 2]
```diff
@@ -5,23 +5,116 @@
...
+
+func (host *Host) toMapStr() common.MapStr {
+	mapstr := common.MapStr{
+		// https://github.com/elastic/ecs#-host-fields
+		"uptime":              host.uptime,
+		"boottime":            host.info.BootTime,
+		"containerized":       host.info.Containerized,
+		"timezone.name":       host.info.Timezone,
+		"timezone.offset.sec": host.info.TimezoneOffsetSec,
+		"name":                host.info.Hostname,
```
Comment (tsg) [inline x-pack/auditbeat/module/system/host/host.go]: `hostname` would fit better here?

File: x-pack/auditbeat/module/system/host/_meta/data.json (hunk 1 above)
Comment (cwurm) [reply x-pack/auditbeat/module/system/host/_meta/data.json]: This comes from the Metricbeat testing framework (`[data_generator.go](https://github.com/elastic/beats/blob/master/metricbeat/mb/testing/data_generator.go#L151)`). So I think it should be different in 6.6.

File: x-pack/auditbeat/module/system/host/host.go (hunk 2 above)
Comment (cwurm) [reply x-pack/auditbeat/module/system/host/host.go]: 👍

File: x-pack/auditbeat/module/system/host/host.go [hunk 3]
```diff
@@ -5,23 +5,116 @@
...
+
+	// We detect changes to the ID separately, so ignore here.
+	mapstr.Delete("id")
+
+	// Changes in uptime and boottime indicate a system reboot,
+	// which we report separately.
+	mapstr.Delete("uptime")
+	mapstr.Delete("boottime")
+
+	h.WriteString(mapstr.String())
```
Comment (andrewkroh) [inline x-pack/auditbeat/module/system/host/host.go]: The combination of creating a map and then calling `MapStr.String()` which IIRC marshals the data as JSON will be a fairly expensive operation. I think it would be better to directly write the values that should be included in the hash calculation into the `Hash` via its `io.Writer` interface.

File: x-pack/auditbeat/module/system/host/host.go [hunk 4]
```diff
@@ -31,123 +124,251 @@ func init() {
...
+
+		eventNewHost := hostEvent(currentHost, eventTypeEvent, eventActionIdChanged)
+		eventNewHost.MetricSetFields.Put("old_id", ms.lastHost.info.UniqueID)
+		events = append(events, eventNewHost)
+	}
+
+	// Report reboots separately
+	if !currentHost.info.BootTime.Equal(ms.lastHost.info.BootTime) {
+		events = append(events, hostEvent(currentHost, eventTypeEvent, eventActionReboot))
+
```
Comment (andrewkroh) [inline x-pack/auditbeat/module/system/host/host.go]: There’s an extra newline here.

Comment (cwurm) [reply x-pack/auditbeat/module/system/host/host.go]: 👍

File: x-pack/auditbeat/module/system/host/host.go (hunk 3 above)
Comment (cwurm) [reply x-pack/auditbeat/module/system/host/host.go]: You're right, I've changed to using the fields directly.

File: x-pack/auditbeat/module/system/host/host.go [hunk 5]
```diff
@@ -5,23 +5,116 @@
...
 	moduleName    = "system"
 	metricsetName = "host"
+
+	bucketName        = "host.v1"
+	bucketKeyLastHost = "lastHost"
+
+	eventTypeState = "state"
+	eventTypeEvent = "event"
+
+	eventActionHost        = "host"
```
[the remark this answers is not in the document]
Comment (cwurm) [reply x-pack/auditbeat/module/system/host/host.go]: done

File: x-pack/auditbeat/module/system/host/config.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+// config defines the metricset's configuration options.
+type config struct {
+	StatePeriod     time.Duration `config:"state.period"`
+	HostStatePeriod time.Duration `config:"host.state.period"`
+}
+
+// validate validates the metricset config.
+func (c *config) validate() error {
```
[the remark this answers is not in the document]
Comment (cwurm) [reply x-pack/auditbeat/module/system/host/config.go]: Ah, makes sense. I removed it.

File: x-pack/auditbeat/module/system/host/host.go [hunk 7]
```diff
@@ -55,17 +78,19 @@ type Host struct {
...
+	if host.info.Containerized != nil {
+		h.WriteString(strconv.FormatBool(*host.info.Containerized))
+	}
 
-	// Changes in uptime and boottime indicate a system reboot,
-	// which we report separately.
-	mapstr.Delete("uptime")
-	mapstr.Delete("boottime")
+	h.WriteString(host.info.Timezone)
+	h.WriteString(strconv.Itoa(host.info.TimezoneOffsetSec))
```
Comment (andrewkroh) [inline x-pack/auditbeat/module/system/host/host.go]: Rather than converting the number to a string (which requires allocation) I think you can use

`binary.Write(h, binary.BigEndian, int32(host.Info.TimezoneOffsetSec))`

which will write the bytes directly into the hash. It should not fallback to reflection if you cast from `int` to `int32`.

Comment (cwurm) [reply x-pack/auditbeat/module/system/host/host.go]: Makes sense. Does it matter if it writes big-endian or little-endian? I see us using both quite heavily in the Beats code.

Comment (andrewkroh) [reply x-pack/auditbeat/module/system/host/host.go]: Endian doesn’t matter. It only matters that it’s hard-coded so that the hash doesn’t change between architectures.

File: x-pack/auditbeat/module/system/host/host.go [hunk 8]
```diff
@@ -5,23 +5,145 @@
...
+func (host *Host) toMapStr() common.MapStr {
+	mapstr := common.MapStr{
+		// https://github.com/elastic/ecs#-host-fields
+		"uptime":              host.uptime,
+		"boottime":            host.info.BootTime,
+		"timezone.name":       host.info.Timezone,
+		"timezone.offset.sec": host.info.TimezoneOffsetSec,
+		"hostname":            host.info.Hostname,
+		"id":                  host.info.UniqueID,
+		// TODO "host.type": ?
```
Comment (andrewkroh) [inline x-pack/auditbeat/module/system/host/host.go]: Can you put some more details about what needs to be done here with this TODO. And possibly link it back to some issue if there is some that tracks the remaining work. Or remove it.

Comment (cwurm) [reply x-pack/auditbeat/module/system/host/host.go]: It referred to implementing the ECS field `host.type` - but I don't see what with so I'm removing it.

File: x-pack/auditbeat/module/system/host/host.go [hunk 9]
```diff
@@ -5,23 +5,144 @@
...
-
 	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
 	"github.com/elastic/go-sysinfo"
+	"github.com/elastic/go-sysinfo/types"
 )
 
 const (
 	moduleName    = "system"
 	metricsetName = "host"
```
Comment (ruflin) [inline x-pack/auditbeat/module/system/host/host.go]: @andrewkroh @tsg Should this be `audit.host`?

Comment (cwurm) [reply x-pack/auditbeat/module/system/host/host.go]: I'm planning to rename the module in a separate PR.

Comment (cwurm) [reply x-pack/auditbeat/module/system/host/host.go]: I was planning to namespace the fields only:

```
func init() {
	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
		mb.DefaultMetricSet(),
        mb.WithNamespace("system.audit")
	)
}
```

Comment (ruflin) [reply x-pack/auditbeat/module/system/host/host.go]: With https://github.com/elastic/beats/pull/9393 it will become important that `event.dataset` is unique, meaning it should be `system.audit.process` here. Let's see what tricks we can apply here.

Comment (ruflin) [reply x-pack/auditbeat/module/system/host/host.go]: +1 on getting the PR in and then as soon as #9393 is also merged, will figure it out.

Comment (cwurm) [reply x-pack/auditbeat/module/system/host/host.go]: I see how having data from two different metricsets having the same `event.dataset` can be confusing. But I don't think that is different from having the same `event.module` and `event.metricset`.

Still, I could see value in having unique dataset names. Who knows, maybe sometime in the future we might have only datasets, and no more modules/metricsets.

I opened https://github.com/elastic/beats/pull/9499 for namespacing the system module, we can discuss any changes there.
